groups/nts/ntsu/ntsu_socketutil.cpp

```diff
@@ -539,34 +542,180 @@ void IncomingDataStats::update(const struct iovec* buffersReceivable,
-                            bsls::AlignmentFromType<cmsghdr>::VALUE>
-    CtrlMsgBuf;
-
-/// Extract timestamps from control block of specified 'msg'.  Timestamps
-/// are extracted from the first control message of suitable time and saved
-/// into specified 'ctx'.  TimestampingData type contains an array of three
-/// elements, each one represents a timestamp.  Only the first and the last
-/// elements are processed as the middle one is filled only when
-/// SOF_TIMESTAMPING_SYS_HARDWARE flag is set, but this flag is marked
-/// deprecated and ignored
-/// https://docs.kernel.org/networking/timestamping.html
-void extractTimestampsFromCtrlBlock(ntsa::ReceiveContext* ctx,
-                                    const msghdr&         msg)
+    if (options.foreignHandle().isNull()) {
+        return ntsa::Error();
+    }
+
+    msg->msg_control    = d_arena.buffer();
+    msg->msg_controllen = static_cast<socklen_t>(k_SEND_CONTROL_BUFFER_SIZE);
+    
+    int foreignHandle = options.foreignHandle().value();
+
+    struct cmsghdr *ctl = CMSG_FIRSTHDR(msg);
+
+    ctl->cmsg_level = SOL_SOCKET;
+    ctl->cmsg_type  = SCM_RIGHTS;
+    ctl->cmsg_len   = CMSG_LEN(sizeof(int));
+
+    bsl::memcpy(CMSG_DATA(ctl), &foreignHandle, sizeof(int));
+    
+    return ntsa::Error();
+}
+
+NTSCFG_INLINE
+ReceiveControl::ReceiveControl()
+{
+}
+    
+NTSCFG_INLINE
+ReceiveControl::~ReceiveControl()
+{
+}
+
+void ReceiveControl::initialize(msghdr* msg)
+{
+    bsl::memset(d_arena.buffer(), 0, k_RECEIVE_CONTROL_BUFFER_SIZE);
+
+    msg->msg_control    = d_arena.buffer();
+    msg->msg_controllen = 
+        static_cast<socklen_t>(k_RECEIVE_CONTROL_BUFFER_SIZE);
+}
+
+ntsa::Error ReceiveControl::decode(ntsa::ReceiveContext*       context, 
+                                   const msghdr&               msg,
+                                   const ntsa::ReceiveOptions& options)
 {
     for (cmsghdr* hdr = CMSG_FIRSTHDR(&msg); hdr != 0;
          hdr          = CMSG_NXTHDR(const_cast<msghdr*>(&msg), hdr))
     {
         if (hdr->cmsg_level == SOL_SOCKET) {
-            if (hdr->cmsg_type == TimestampUtil::e_SCM_TIMESTAMPING) {
+            if (hdr->cmsg_type == SCM_RIGHTS) {
+                int foreignHandle = -1;
+
+                if (hdr->cmsg_len != CMSG_LEN(sizeof(int))) {
+                    BSLS_LOG_WARN("Ignoring received control block meta-data: "
+                                  "Unexpected control message payload size: "
+                                  "expected %d bytes, found %d bytes", 
+                                  (int)(CMSG_LEN(sizeof(int))), 
+                                  (int)(hdr->cmsg_len));
+                    continue;
+                }
+
+                bsl::memcpy(&foreignHandle, CMSG_DATA(hdr), sizeof(int));
+
+                if (options.wantForeignHandles()) {
+                    context->setForeignHandle(foreignHandle);
+                }
+                else {
+                    ::close(foreignHandle);
```

Comment:
Maybe we should have here WARN of DEBUG level print? Otherwise there is no way at all to know that a foreigh process wanted to send us a file descriptor while we were not interested in that.

Reply:
Done.